### ISSUE TITLE
Replace all uses of MatElm and SetMatElm

### ISSUE THREAD

**Issue by fingolfin** [homalg-project/homalg_project#271]
Please consider replacing all uses of `MatElm` and `SetMatElm` in the packages in this repository (and also any other HomAlg package) by their modern counterparts (fully supported since GAP 4.9). That is:
- replace any occurrence of `MatElm(mat, row, col)` by `mat[row, col]`
- replace any occurrence of `SetMatElm(mat, row, col, val);` by `mat[row, col] := val;`
- replace method installations for `MatElm` by method installations for `\[\]` resp. `ELM_LIST` (both reference the same operation)
- replace method installations for `SetMatElm` by method installations for `\[\]\:\=` resp. `ASS_LIST` (both reference the same operation)

**Comment by mohamed-barakat** [comment]
Thanks, I will.

(I used `MatElm` and `SetMatElm` as they were part of the `MatrixObj` interface).

**Comment by fingolfin** [comment]
Yeah, to clarify: I didn't meant to suggest here that homalg did anything wrong -- it most definitely did not!

It's just that we have something better than `MatElm`/`SetMatElm` now. But you can actually keep using them for the foreseeable future, as we'll return them for backwards compatibility. It just would be nice to gradually phase out all of its uses, that's all. Plus I think the resulting code will look nicer / be easier to read (though I guess that's highly subjective and a matter of taste).

**Comment by mohamed-barakat** [comment]
I fully agree, the readability argument alone convinces me :)

**Comment by fingolfin** [comment]
Hmm, also, I just discovered that homalg overloads `MatElm` and `SetMatElm` with variants that take one extra argument:
```gap
DeclareOperation( "SetMatElm",
        [ IsHomalgMatrix, IsInt, IsInt, IsString, IsHomalgRing ] );

DeclareOperation( "SetMatElm",
        [ IsHomalgMatrix, IsInt, IsInt, IsRingElement, IsHomalgRing ] );
...
DeclareOperation( "MatElm",
        [ IsHomalgMatrix, IsInt, IsInt, IsHomalgRing ] );
```

That code then will have to keep using `MatElm` and `SetMatElm`. But that shouldn't be a major drawback, I guess?

In the meantime, if desired, I can provide a pull request that implements this partially, created with some careful regex based search-and-replace

**Comment by mohamed-barakat** [comment]
> Hmm, also, I just discovered that homalg overloads `MatElm` and `SetMatElm` with variants that take one extra argument:
> 
> ```gap
> DeclareOperation( "SetMatElm",
>         [ IsHomalgMatrix, IsInt, IsInt, IsString, IsHomalgRing ] );
> 
> DeclareOperation( "SetMatElm",
>         [ IsHomalgMatrix, IsInt, IsInt, IsRingElement, IsHomalgRing ] );
> ...
> DeclareOperation( "MatElm",
>         [ IsHomalgMatrix, IsInt, IsInt, IsHomalgRing ] );
> ```
> 
> That code then will have to keep using `MatElm` and `SetMatElm`. But that shouldn't be a major drawback, I guess?

No, it won't be a problem.

> In the meantime, if desired, I can provide a pull request that implements this partially, created with some careful regex based search-and-replace

Oh, this is a lovely offer and hard to reject :)

**Comment by fingolfin** [comment]
Small update: I changed my mind and now would actually suggest to leave the `InstallMethod` invocations as they are right now. I may change some things related to this in GAP soon, and it'll be less work if you just leave those as they are.

In the meantime, I submitted a bunch of PRs to replace many uses of `MatElm` and `SetMatElm`. This leaves a few of them which use the additional ring argument mentioned above (but this is really just a very small number). But I also don't touch a couple uses like this:  `MatElm( R * w , 1, 1 );` which technically could be rewritten as `(R * w)[1,1]`, but I have mixed feelings about the readability of this, and thus leave it up to you guys to make that change, or perhaps instead first assign `R * w` to a temp variable. Whatever you prefer.
